libparsec/crates/client/src/client/recovery_device.rs

```diff
@@ -0,0 +1,373 @@
+            PlatformImportRecoveryDeviceError::InvalidPath(error) => {
+                ImportRecoveryDeviceError::InvalidPath(error)
+            }
+            PlatformImportRecoveryDeviceError::InvalidData => {
+                ImportRecoveryDeviceError::InvalidData
+            }
+            PlatformImportRecoveryDeviceError::InvalidPassphrase => {
+                ImportRecoveryDeviceError::InvalidPassphrase
+            }
+            PlatformImportRecoveryDeviceError::DecryptionFailed => {
+                ImportRecoveryDeviceError::DecryptionFailed
+            }
+            PlatformImportRecoveryDeviceError::SaveDeviceError(save_device_error) => {
+                ImportRecoveryDeviceError::SaveDeviceError(save_device_error)
+            }
+        }
+    }
+}
+
+pub async fn register_new_device(
+    cmds: Arc<AuthenticatedCmds>,
+    new_device: LocalDevice,
+    author: Arc<LocalDevice>,
+) -> Result<CertificateBasedActionOutcome, CertifRegisterNewDeviceError> {
+    // Loop is needed to deal with server requiring greater timestamp
+    let mut timestamp = author.now();
+
+    loop {
+        let outcome = internal_register_new_device(&cmds, &new_device, &author, timestamp).await?;
+
+        match outcome {
+            DeviceInternalsOutcome::Done(outcome) => return Ok(outcome),
+            DeviceInternalsOutcome::RequireGreaterTimestamp(strictly_greater_than) => {
+                // TODO: handle `strictly_greater_than` out of the client ballpark by
+                // returning an error
+                timestamp = greater_timestamp(
+                    &author.time_provider,
+                    GreaterTimestampOffset::User,
+                    strictly_greater_than,
+                );
+            }
+        }
+    }
+}
+
+pub(crate) struct DeviceCertificatesBytes {
+    pub full: Bytes,
+    pub redacted: Bytes,
+}
+
+/// generates certificates for new device signed by author at now
+pub(crate) fn generate_new_device_certificates(
+    new_device: &LocalDevice,
+    author: &Arc<LocalDevice>,
+    now: DateTime,
+) -> DeviceCertificatesBytes {
+    let device_cert = DeviceCertificate {
+        author: CertificateSignerOwned::User(author.device_id),
+        timestamp: now,
+        user_id: new_device.user_id,
+        device_id: new_device.device_id,
+        device_label: MaybeRedacted::Real(new_device.device_label.clone()),
+        verify_key: new_device.verify_key(),
+        algorithm: SigningKeyAlgorithm::Ed25519,
+    };
+
+    let device_certificate = device_cert.dump_and_sign(&author.signing_key).into();
+
+    let redacted_device_cert = device_cert.into_redacted();
+
+    let redacted_device_certificate = redacted_device_cert
+        .dump_and_sign(&author.signing_key)
+        .into();
+
+    DeviceCertificatesBytes {
+        full: device_certificate,
+        redacted: redacted_device_certificate,
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum CertifRegisterNewDeviceError {
```

Comment:
should be called `Certif...` since it's not something from the `client/src/certf`